### PR TITLE
Refpolicy aos vm

### PR DESCRIPTION
**Pull request by MykolaSuperman**

### CODE REVIEW

File: README [hunk 1]
```diff
@@ -274,3 +274,68 @@ refresh			Attempts to reinsert all modules that are currently
 xml			Build a policy.xml from the XML included with the
 			base policy headers and any XML in the modules in
 			the current directory.
+
+
+## How to work with SELinux
+### Introduction
+
+To be able to add and change the SELinux policy, you need to clone the repository with refpolicy, and make changes with a separate commit.
```
Comment (Eabramov84) [inline README]: Please write instructions in indirect form(withou you)
e.g. 
To add and change the SELinux policy, the repository with refpolicy should be cloned and changes are done with a separate commit.

File: README [hunk 2]
```diff
@@ -274,3 +274,68 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+Each process or resource (file, dir etc.) has a label (type) that describes what can be accessed.
+
+SELinux considers subject-access-object rule set. On a Linux system, subjects are processes, and object is the resource on which an action applies.
+
+Refpolicy contains the following file types:
+1. **te** - the file contains rules describing what the domain can do (labels of the subjects or objects). This file does not say that this process can run another file, it just says that the process labeled A can read the file labeled B.
+2. **fc** - describes the security context (labels) on a file, directory, etc., which will be applied when the policy is install. By default, directories or files inherit the context of their parents, but you can manually set the desired context through this file
```
Comment (Eabramov84) [inline README]: but the desired context could be manually set using this file

File: README [hunk 3]
```diff
@@ -274,3 +274,68 @@ refresh			Attempts to reinsert all modules that are currently
...
+3. **if** - the interface file creates the macros that other modules will use to gain access to my resources.
```
Comment (Eabramov84) [inline README]: to the resources

File: README [hunk 4]
```diff
@@ -274,3 +274,68 @@ refresh			Attempts to reinsert all modules that are currently
...
+- **scontext** - SELinux context of the process (source) that attempted the denied action.
+- **tcontext** -SELinux context of the object (target) the process attempted to access
+- **comm** - name of the command that was used to access the target context
+name
+- **tclass** - target object class
+- **dev** - device on which the target resides
+- **name** - name of the target
+- can contained **path** -  path to the object (target) the process attempted to access.
+
+Next we need to analyze the received log and find the corresponding **te** file by **scontext**, for example, for **dnsmasq_t** there is a corresponding **dnsmasq.te** file and in this file you need to add a rule that allows **scontext** to access **tcontext**.
```
Comment (Eabramov84) [inline README]: The next step is to analyze ...  **dnsmasq.te** file and in this file  rule that allows **scontext** to access **tcontext** should be added.

File: policy/modules/kernel/files.if [hunk 5]
```diff
@@ -5834,6 +5834,7 @@ interface(`files_dontaudit_rw_var_files',`
 	dontaudit $1 var_t:file rw_file_perms;
 ')
 
+
```
Comment (al1img) [inline policy/modules/kernel/files.if]: Not relevant change.

File: policy/modules/kernel/files.if [hunk 6]
```diff
@@ -6047,6 +6048,24 @@ interface(`files_manage_var_lib_dirs',`
 	allow $1 var_lib_t:dir manage_dir_perms;
 ')
 
+########################################
```
Comment (al1img) [inline policy/modules/kernel/files.if:6157]: Would be nice to have it in separate commit:
* provide access for /var/run
* provide access for efivars
In this case, it would be easier to apply policies for new platform. For example for u-boot devices we should remove efivar access e.g.

File: README [hunk 7]
```diff
@@ -274,3 +274,68 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+
+## How to work with SELinux
+### Introduction
+
+To be able to add and change the SELinux policy, you need to clone the repository with refpolicy, and make changes with a separate commit.
+
+Each process or resource (file, dir etc.) has a label (type) that describes what can be accessed.
+
+SELinux considers subject-access-object rule set. On a Linux system, subjects are processes, and object is the resource on which an action applies.
```
Comment (al1img) [inline README]: Could you use max 120 length line?

Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 1 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 2 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 3 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 4 above)
Comment (MykolaSuperman) [reply README]: Done

File: policy/modules/kernel/files.if (hunk 5 above)
Comment (MykolaSuperman) [reply policy/modules/kernel/files.if]: Done

File: policy/modules/kernel/files.if (hunk 6 above)
Comment (MykolaSuperman) [reply policy/modules/kernel/files.if:6157]: Done. Split to several commits

File: README [hunk 8]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
 xml			Build a policy.xml from the XML included with the
 			base policy headers and any XML in the modules in
 			the current directory.
+
+
+## How to work with SELinux
+### Introduction
+
+To add and change the SELinux policy, the repository with refpolicy, and make changes with
```
Comment (al1img) [inline README]: Looks like something is missing in this sentence.

File: README [hunk 9]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+a separate commit.
+
+Each process or resource (file, dir etc.) has a label (type) that describes what can be accessed.
+
+SELinux considers subject-access-object rule set. On a Linux system, subjects are processes, and object is the resource
+on which an action applies.
```
Comment (al1img) [inline README]: is applied

File: README [hunk 10]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+Refpolicy contains the following file types:
+1. **te** - the file contains rules describing what the domain can do (labels of the subjects or objects). This file
+does not say that this process can run another file, it just says that the process labeled A can read the file
+labeled B.
+2. **fc** - describes the security context (labels) on a file, directory, etc., which will be applied when the policy
+is install. By default, directories or files inherit the context of their parents, but the desired context could
```
Comment (al1img) [inline README]: is installed

File: README [hunk 11]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+To add and change the SELinux policy, the repository with refpolicy, and make changes with
+a separate commit.
+
+Each process or resource (file, dir etc.) has a label (type) that describes what can be accessed.
+
+SELinux considers subject-access-object rule set. On a Linux system, subjects are processes, and object is the resource
+on which an action applies.
+
+Refpolicy contains the following file types:
+1. **te** - the file contains rules describing what the domain can do (labels of the subjects or objects). This file
```
Comment (al1img) [inline README]: Add full name for each type: `te - (type enforcement) `

File: README [hunk 12]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+does not say that this process can run another file, it just says that the process labeled A can read the file
+labeled B.
+2. **fc** - describes the security context (labels) on a file, directory, etc., which will be applied when the policy
```
Comment (al1img) [inline README]: `fc - (file context)`

File: README [hunk 13]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+is install. By default, directories or files inherit the context of their parents, but the desired context could
+be manually set using this file.
+3. **if** - the interface file creates the macros that other modules will use to gain access to the resources.
```
Comment (al1img) [inline README]: `if - (...)`

File: README [hunk 14]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+In the **permissive** mode, SELinux is active, the security policy is loaded, the file system is labeled and access
+denial entries are logged. However, the policy is not enforced and thus no access is actually denied.
+In the **enforced** mode, the security policy is applied. Each access that is not explicitly allowed by the policy
+is denied.
+
+To be able to check SELinux for possible access denials, need to set the **permissive** mode in the **meta-aos-vm**
+for **refpolicy-aos** recipe to the **DEFAULT_ENFORCING** variable. This will make it possible to use SELinux,
+but in logging mode.
+
+To diagnose denials security policies, using the following command `journalctl | grep avc`, if the log contains
```
Comment (al1img) [inline README]: using -> use

File: README [hunk 15]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+**avc: denied** that means it is an SELinux policy denial. This log will include a line like this:
+
+```
+AVC avc:  denied  { write } for  pid=1248 comm="dnsmasq" name="dev-log" dev="tmpfs" ino=20945
+scontext=system_u:system_r:dnsmasq_t:s0 tcontext=system_u:object_r:syslogd_runtime_t:s0 tclass=sock_file permissive=0
+```
+
+- **scontext** - SELinux context of the process (source) that attempted the denied action.
+- **tcontext** -SELinux context of the object (target) the process attempted to access
```
Comment (al1img) [inline README]: missing space after `-`

File: README [hunk 16]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+To diagnose denials security policies, using the following command `journalctl | grep avc`, if the log contains
+**avc: denied** that means it is an SELinux policy denial. This log will include a line like this:
+
+```
+AVC avc:  denied  { write } for  pid=1248 comm="dnsmasq" name="dev-log" dev="tmpfs" ino=20945
+scontext=system_u:system_r:dnsmasq_t:s0 tcontext=system_u:object_r:syslogd_runtime_t:s0 tclass=sock_file permissive=0
+```
+
+- **scontext** - SELinux context of the process (source) that attempted the denied action.
```
Comment (al1img) [inline README]: remove period at the end of this item

File: README [hunk 17]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+- **tcontext** -SELinux context of the object (target) the process attempted to access
+- **comm** - name of the command that was used to access the target context name
+- **tclass** - target object class
+- **dev** - device on which the target resides
+- **name** - name of the target
+- can contained **path** -  path to the object (target) the process attempted to access.
+
+The next step is to analyze the received log and find the corresponding **te** file by **scontext**, for example,
+for **dnsmasq_t** there is a corresponding **dnsmasq.te** file and in this file need to add a rule that
```
Comment (al1img) [inline README]: in this file a rule that allows ... has to be added.

File: README [hunk 18]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+allows **scontext** to access **tcontext**.
+
+### Recommended steps for editing refpolicy.
```
Comment (al1img) [inline README]: Remove period at the end

File: README [hunk 19]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+To find the required **te** file where to add the rule need to find a line declaring the **scontext** type
```
Comment (al1img) [inline README]: Here and in the following sentences, it looks like the main noun is missed. Use either imperative: `To find the required **te** file where to add the rule, find a line declaring ...` or `you need to find...` or passive voice: `To find the required **te** file where to add the rule, a line declaring ... has to be found`.

File: README [hunk 20]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+(label) how in example `type dnsmasq_t;`. Next, after the **Local policy comment**, add the following rule.
```
Comment (al1img) [inline README]: 1. how in example -> as in example
2. `add the following rule` - which one?

File: README [hunk 21]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+Sometimes the rules are logically grouped, for example, there are rules for working with the `/etc` directory, etc.
+
+**It is important** that before adding a rule, need to look for a possible macro for **tcontext** declared in a
+file with the **if** extension, to find the necessary file **if**, first need to find the corresponding **te** file
```
Comment (al1img) [inline README]: Start new sentence with `to find necessary...`

File: README [hunk 22]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+in which the label will be declared and then take the **if** file with the same name. For example, the
+`syslogd_runtime_t` label is declared in the `logging.te` file, then the macro should be looked for in the
+`logging.if` file.
+
+If some file or directory marked with an incorrect label or `unlabeled_t`, then  need to find the **te**
```
Comment (al1img) [inline README]: is marked

File: README [hunk 23]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+file where the label of interest is declared and take the file with the **fc** extension and declare the context for
+the label of interest there. For example
+```
```
Comment (al1img) [inline README:344]: Add empty line before and after code snippets. I suggest to install markdown linter for vscode.

File: README [hunk 24]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+**It is important** that before adding a rule, need to look for a possible macro for **tcontext** declared in a
+file with the **if** extension, to find the necessary file **if**, first need to find the corresponding **te** file
+in which the label will be declared and then take the **if** file with the same name. For example, the
+`syslogd_runtime_t` label is declared in the `logging.te` file, then the macro should be looked for in the
+`logging.if` file.
+
+If some file or directory marked with an incorrect label or `unlabeled_t`, then  need to find the **te**
+file where the label of interest is declared and take the file with the **fc** extension and declare the context for
+the label of interest there. For example
```
Comment (al1img) [inline README]: `For example:`

File: README [hunk 25]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+```
+/usr/bin/aos_vis  --  gen_context(system_u:object_r:aos_exec_t,s0)
+```
+which means that the `aos_vis` file will be labeled `aos_exec_t`, this is an executable process.
+
+To see what would be the correct or possible label in the `unlabeled_t`, need to run the `fixfiles -F -f relabel`
+command or another command, after that the resource or process's will be labeled with proper context.
```
Comment (al1img) [inline README]: 1. which another command?
2. Start new sentence with `after`.

File: README [hunk 26]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+**It is important** if need to create a new label, then must be create three files **te, fc, if** with the name
```
Comment (al1img) [inline README]: `If a new label needs to be created, it is important to create three files....` or `create three files...` `with the name of the label`.

File: README [hunk 27]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+of the label that with which is created. At the same time, the **te** file in which declare a new label type must not
```
Comment (al1img) [inline README]: The **te** file that declares (or in which the new label is declared) should not be empty whereas **fc, if** files can be empty,

File: README [hunk 28]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+be empty, the **fc, if** files can be empty.
+
+Sometimes in the logs contains denials security policies, but which may not be very critical, for example, get the
```
Comment (al1img) [inline README]: 1. `Sometimes, logs may contain denials security policies that are not very critical, for example, getting... or reading`
2. Need to be restructured: for example, ... , for example.

File: README [hunk 29]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+file attribute or read a file that may not affect the system, for example, when login into the system, being created
+policies to denial write to the system log
```
Comment (al1img) [inline README]: log:

File: README [hunk 30]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+```
+avc:  denied  { write } for  pid=1034 comm="login" name="dev-log" dev="tmpfs" ino=20945
+scontext=system_u:system_r:local_login_t:s0-s0:c0.c1023
+tcontext=system_u:object_r:syslogd_runtime_t:s0 tclass=sock_file permissive=0
+```
+For these denials **not need to add rules** to the refpolicy if it is not necessary.
```
Comment (al1img) [inline README]: `no need to add rules` or `rules may not be added`

File: README [hunk 31]
```diff
@@ -274,3 +274,101 @@ refresh			Attempts to reinsert all modules that are currently
...
+
+After the corresponding security policy is added, it is necessary to build the refpolicy again through yocto.
+To find interest a custom module with the **pp** extension and install it on the system without reinstalling
```
Comment (al1img) [inline README]: Unclear sentence, please rephrase.

File: README (hunk 9 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 10 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 11 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 12 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 13 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 14 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 15 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 16 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 17 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 18 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 20 above)
Comment (MykolaSuperman) [reply README]: 1. Done
2. use `desired rule`

File: README (hunk 21 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 22 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 24 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 29 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 27 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 26 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 23 above)
Comment (MykolaSuperman) [reply README:344]: Done

File: README (hunk 19 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 25 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 28 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 8 above)
Comment (MykolaSuperman) [reply README]: Remove this sentence

File: README (hunk 30 above)
Comment (MykolaSuperman) [reply README]: Done

File: README (hunk 31 above)
Comment (MykolaSuperman) [reply README]: Removed this sentence as it is incorrect in our context

File: README [hunk 32]
```diff
@@ -274,3 +274,100 @@ refresh			Attempts to reinsert all modules that are currently
...
+```
+
+which means that the `aos_vis` file will be labeled `aos_exec_t`, this is an executable process.
+
+To see what would be the correct or possible label in the `unlabeled_t`, run the `fixfiles -F -f relabel`
+command or another command `restorecon`. After that the resource or process's will be labeled with proper context.
+
+**It is important** If a new label needs to be created, it is important to create three files **te, fc, if** with the name
+of the label that with which is created. The **te** file that declares (or in which the new label is declared) should not
+be empty whereas **fc**, **if** files can be empty
```
Comment (al1img) [inline README]: Period at the end.

Comment (MykolaSuperman) [reply README]: Done

**Review by al1img — APPROVED** [review]
